### DISCUSSION TITLE
Detecting changes in arrow "Datasets" in a targets pipeline

### DISCUSSION THREAD

**Discussion by boshek** [ropensci/targets#551] [Help] [answered]
Apologies if this is answered very clearly in the documentation somewhere that I haven't found yet. 

I work with [arrow datasets](https://arrow.apache.org/docs/r/articles/dataset.html) quite a bit. The structure of this data is such that, the function `arrow::open_dataset` points at the top level directory not a file. Occasionally I have scenarios where that data is provisioned outside of my control (i.e. will just change). Is there a way in a targets pipeline to detect whether that data has changed and re-run that particular bit of code? Here is a reprex that updates the data between pipeline runs but targets isn't recognizing that the data has changed:

``` r
library(dplyr, warn.conflicts = FALSE)
library(arrow, warn.conflicts = FALSE)
library(targets, warn.conflicts = FALSE)

## Create original data
mtcars %>%
  filter(mpg > 15) %>%
  group_by(cyl) %>%
  write_dataset("data/")


tar_script({
  library(arrow, warn.conflicts = FALSE)
  library(dplyr, warn.conflicts = FALSE)
  list(
    tar_target(data_path, "data"),
    tar_target(dataset, open_dataset(data_path)),
    tar_target(dataset_in_mem, nrow(dataset))
  )
})

tar_make()
#> * start target data_path
#> * built target data_path
#> * start target dataset
#> * built target dataset
#> * start target dataset_in_mem
#> * built target dataset_in_mem
#> * end pipeline

original <- tar_read(dataset_in_mem)


## Now a change is made to the data
mtcars %>%
  group_by(cyl) %>%
  write_dataset("data/")

tar_make()
#> v skip target data_path
#> v skip target dataset
#> v skip target dataset_in_mem
#> v skip pipeline

changed <- tar_read(dataset_in_mem)

## They are the same.
original
#> [1] 26
changed
#> [1] 26
```

One thought I had is to _always_ run the `dataset` target but that seems a bit heavy handed. Any thoughts on how to proceed?

**Comment by wlandau** [comment, marked as answer]
`targets` can track external files and directories with `tar_target(format = "file")`: https://books.ropensci.org/targets/files.html#external-input-files. So in your target list, you would write `tar_target(data_path, "data", format = "file")`. (Related: if you write in `arrow` formats feather or Parquet, there is also `format = "feather"` and `format = "parquet"`.)

**Reply by boshek** [reply]
Thank you for the response. I am very suspicious of myself here but I _think_ that maybe this isn't working. Specifying `format` doesn't appear to propagate down to `dataset_in_mem`. Or am I just specifying it wrong?

``` r
library(dplyr, warn.conflicts = FALSE)
library(arrow, warn.conflicts = FALSE)
library(targets, warn.conflicts = FALSE)

## Create original data
mtcars %>%
  filter(mpg > 15) %>%
  group_by(cyl) %>%
  write_dataset("data/")


tar_script({
  library(arrow, warn.conflicts = FALSE)
  library(dplyr, warn.conflicts = FALSE)
  list(
    tar_target(data_path, "data", format = "file"),
    tar_target(dataset, open_dataset(data_path)),
    tar_target(dataset_in_mem, nrow(collect(dataset)))
  )
})

tar_make()
#> * start target data_path
#> * built target data_path
#> * start target dataset
#> * built target dataset
#> * start target dataset_in_mem
#> * built target dataset_in_mem
#> * end pipeline

original <- tar_read(dataset_in_mem)


## Now a change is made to the data
mtcars %>%
  group_by(cyl) %>%
  write_dataset("data/")

tar_make()
#> * start target data_path
#> * built target data_path
#> * start target dataset
#> * built target dataset
#> v skip target dataset_in_mem
#> * end pipeline

changed <- tar_read(dataset_in_mem)

## They are the same.
original
#> [1] 26
changed
#> [1] 26
```

**Reply by wlandau** [reply]
`open_dataset()` returns a pointer (R6 class, really) of a data object, not the data itself. Apparently this object has the same hash no matter what is in the actual data. I recommend returning a simpler object that reflects what is going on in the data, ideally the dataset itself. The concerns are similar to https://books.ropensci.org/targets/practices.html#return-value.

**Reply by boshek** [reply]
Ok I see now. That makes sense.

**Reply by nealrichardson** [reply]
I'm not familiar with target's internals, but for what it's worth, R6 objects (like arrow's `Dataset`) are `environment`s, and it appears that different environments may hash the same:

```
> e1 <- new.env()
> e2 <- new.env()
> e1
<environment: 0x1289030c8>
> e2
<environment: 0x128902a00>
> identical(e1, e2)
[1] FALSE

> digest::digest(e1)
[1] "391768beb011f250a0024d09749ca504"
> digest::digest(e2)
[1] "391768beb011f250a0024d09749ca504"
> identical(digest::digest(e1), digest::digest(e2))
[1] TRUE
```

**Reply by wlandau** [reply]
If the `arrow` `R6` object had the actual data in it, that would help, but it looks you need `collect()` to retrieve it. But even if it did have the data, `R6` classes are brittle to hash because they contain functions: https://github.com/ropensci/drake/issues/345. For ordinary functions, `targets` deparses the function before hashing. But for `R6` classes and reference classes, `targets` does not try to do anything special.

**Comment by rsangole** [comment]
@boshek (and @wlandau),

Did you happen to find a solution to working with arrow `open_dataset()` objects with `targets`? I use targets to make production ready pipelines and all my datasets are in partitioned parquet files which I read in with `open_dataset`. I faced the same issues you are today, and found this thread.

**Reply by wlandau** [reply]
You could try defining a custom tar_format() for this purpose. With tar_format(), you can configure how to read, write, marshal, and unmarshal the object.

**Reply by rsangole** [reply]
I see. I'll try reading up on this; I don't know what (un)marshaling means.

**Reply by wlandau** [reply]
See https://docs.ropensci.org/targets/reference/tar_format.html#marshalling and https://cran.r-project.org/web/packages/future/vignettes/future-4-non-exportable-objects.html for details.
